File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
        void f(vector<int> nums, vector<vector<int>> &ans, vector<int> &ds, int i)
        {
             if(i==nums.size())
             {
                    ans.push_back(ds);
                     return;
             }
                ds.push_back(nums[i]);
                f(nums, ans, ds, i+1);
                ds.pop_back();
                f(nums, ans, ds, i+1);
        }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
            vector<int> ds;
            f(nums, ans, ds, 0);
            return ans;
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
         List<List<Integer>>  ans   =  new ArrayList<List<Integer>>();
         helper(nums, ans, 0 , new ArrayList<Integer>());
         return ans;
      }
    private void helper(int[]nums, List<List<Integer>>  ans, int index, List<Integer> subList){
        ans.add(new ArrayList<>(subList));
        for(int i = index;i<nums.length;i++){
          subList.add(nums[i]);
          helper(nums,ans,  i+1, subList);     
          subList.remove(subList.size()-1);
        }
   }
}